[1단계 - 출석 구현] 도기(이동엽) 미션 제출합니다.

dye0p opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
안녕하세요 이프!🖐️ 이번 출석 미션의 리뷰이 도기입니다! 리뷰에 참여해주셔서 감사합니다. 🙇‍♂️
미션을 진행하면서 맞추지 못한 요구사항도 많고, 리팩터링 할 사항도 많지만 미션을 진행하면서 고민했던 부분들을 최대한 말씀드리겠습니다!🔥

## 체크 리스트

- [ ] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [셀프 체크](https://prolog.techcourse.co.kr/studylogs/3961)


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [x] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
TDD를 몇번 시도해봤지만 전부 실패로 돌아갔습니다. 그래서 이번 미션에는 최대한 TDD를 적용해 보았습니다. 사실 TDD로 시도를 하다가 도저히 진도가 안나가서 중간중간 TDD를 안하기도 했지만 도메인에 대한 로직은 최대한 TDD를 적용하려 했습니다! PR을 보내면서 리팩터링 해야할 부분들이 많았지만 시간부족으로 리팩터링을 하지 못했습니다.. 때문에 프로그래밍 요구사항이나 객체지향 생활체조 원칙을 전혀 충족하지 못했다고 생각합니다.🥲

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

## 1. TDD
이번 미션은 TDD로 진행하는 미션이었습니다. 평소 TDD에 관심이 많아, 몇번 시도를 했지만 모두 실패 했습니다.
그래서 이번 미션은 뷰 로직을 제외하고 최대한 TDD로 진행하려고 노력했는데요!

### Top-Down` vs `Bottom-Up
가장 컷던 고민은 테스트를 작성할 때 어떤 크기의 테스트 부터 작성할지 고민이 되었습니다. 큰 단위의 테스트 부터 작성한다면 고려해야할 사항이 많아진다는 단점이 있었고, 작은 단위의 테스트 부터 작성한다면 이것들을 나중에 어떻게 조립할지 고민이었습니다. 이 부분에 대한 이프의 생각은 어떤지 궁금합니다!

### 뷰(view), 컨트롤러(Controller)의 부재로 인한 불안함
TDD는 핵심 도메인 로직의 테스트를 먼저 작성하면서 개발을 하는 것 이라고 알고 있습니다. 하지만 점점 도메인의 규모가 커지고 테스트가 복잡해짐으로 인해서 이러한 테스트가 이후 뷰(view)와 컨트롤러(Controller)에 잘 조합이 될지가 불안요소로 다가왔습니다. 도메인 테스트만 하다보니 콘솔에 직접 입력하여 눈으로 실행결과를 확인하고 싶다는 생각이 많이 들었습니다. 이러한 고민에 대한 이프의 생각은 어떤지 궁금합니다!

### 어떤 객체에 어떤 테스트를 만들어야 하지?
TDD 특성상 먼저 테스트를 만들어야 하는데, 이때 `해당 테스트는 어떤 객체가 가지는 책임이지?` 라는 생각이 들었습니다. 그래서 처음엔 예상하여 테스트를 작성했지만, 점점 시간이 지나고 크기가 커지다보니 원래 작성했던 테스트가 완전히 필요없어져 버리는 경우가 빈번하게 일어났습니다. 코드를 작성할수록 필요한 객체가 많아지게 되고, 필요한 테스트를 하기위한 객체를 판단하기가 어려워졌습니다. 이 부분에 대한 이프의 생각은 어떤지 궁금합니다!

## 2. View 클래스의 책임
`view에서 validation 로직을 작성해줘도 괜찮을까?` 에 대해서 페어와 의논을 했습니다.

- 입력 클래스에서 사용되는 validation 로직은 view에서 작성하지 않고, InputValidator.class 와 같은 유틸성 클래스로 나눈다.
- 입력 클래서에서 사용되는 validation 로직도 비즈니스 로직으로 볼 수 있는가?

위와 마찬가지로 결과값을 출력할 OutputView에서도 파싱이나 변환과 같은 비즈니스 로직을 가지고 있어도 될지에 대한 고민도 들었습니다. 
이 부분에 대한 이프의 생각은 어떤지 궁금합니다!

## 아쉬운 점
1. 시간이 부족해 리팩터링을 못한 부분과 객체지향 생활 체조를 지키지 못한 점, 구현하기에 급급해 객체지향적 설계에 대한 충분한 생각을 하지 못했습니다. 그 결과로 객체지향적이지 못한 코드가 되었다고 생각합니다. 

2. 커밋 사항에서 README 반영을 잘 하지 못한것이 아쉽습니다.

3. 초반 문제에 대한 요구사항을 정확하게 분석하지 않아, 자세한 README를 작성하지 못한것이 아쉽습니다.

4. TDD 적용의 어려움으로 TDD의 장점을 제대로 느끼지 못했습니다. 아직은 TDD가 어떻게 객체지향적인 코드를 작성하는 것인지 제대로 와닫지가 않았습니다. TDD를 잘 하기위한 이프만의 팁이 있다면 알려주세요!🙇‍♂️

이번 리뷰에 참여해주셔서 감사합니다: ) 수고하셨습니다!

## 리팩터링 사항
1. else문을 사용하여 가독성이 저하되었던 문제와 비슷한 모양의 메서드를 하나의 메서드로 추출하여 중복을 제거 하였습니다.
2. Enum에서 정적 팩토리 메서드 사용시 Enum 객체를 리턴하도록 변경하였습니다.

dye0p committed:
docs: 기능 요구사항 추가

dye0p committed:
test: 크루 생성 테스트 추가

dye0p committed:
test: 출석부 생성 테스트 추가

dye0p committed:
feat: attendances 파일 읽어와 출석부 생성기능 추가

dye0p committed:
feat: Calender 클래스 생성

dye0p committed:
feat: 출석 상태 enum 클래스 생성

dye0p committed:
feat: 커스텀 Console 클래스 생성

dye0p committed:
feat: 기능 선택 안내 입력받기 추가

dye0p committed:
docs: 파일 읽어오기 기능 구현 체크

dye0p committed:
remove: 사용하지 않는 클래스 제거

dye0p committed:
feat: 닉네임 입력 기능 추가

dye0p committed:
feat: 출석 기록 확인하기 출력

dye0p committed:
feat: 입력받은 출석 Attendance객체에 저장하는 기능 추가

dye0p committed:
refactor: Crew 객체에 equals() & hashCode() 메서드 추가

dye0p committed:
test: 출석 확인 테스트 생성(월요일/나머지 구분)

dye0p committed:
refactor: findStateBy 메서드 파라미터 타입 변경

- 기존의 요일을 외부에서 가공 후 String으로 받았던 것을 요일의 value 값으로 받아 메서드 안에서 가공하도록 변경

dye0p committed:
feat: 출석 수정과 출력 기능 추가

dye0p committed:
test: Attendance 출석 등록/수정 테스트 추가

dye0p committed:
test: 날짜로부터 요일을 반환하는 Calender 테스트 추가

dye0p committed:
fix: 출석 수정의 출석 상태 계산 매개변수 수정

dye0p committed:
feat: 특정 크루의 출석 기록 확인 기능 추가

dye0p committed:
feat: 크루의 출석 기록 출력 기능 추가

dye0p committed:
feat: 제적 위험자 계산 기능 추가

dye0p committed:
feat: 출석 기록 계산 결과 출력 기능 추가

dye0p committed:
feat: AbsenceResultDto 추가

- AbsenceResultDto를 추가하여 크루 출석 기록 출력 하도록 변경

dye0p committed:
feat: 제적 위험자 결과 조회 출력 기능 추가

dye0p committed:
feat: MainController 추가

dye0p committed:
feat: 등교시간이 캠퍼스 운영시간이 아니라면 예외 추가

dye0p committed:
feat: 공휴일에 출석확인시 예외처리 추가

- 공휴일 출석확인에 대한 성공 & 실패 테스트 추가

dye0p committed:
feat: 이미 출석한 크루가 중복 출석한 경우 예외처리 추가

- 중복 출석 예외처리에 대한 성공 & 실패 테스트 추가

dye0p committed:
refactor: 공휴일 출석 예외메세지 수정

dye0p committed:
feat: 출석 수정이 공휴일인 경우 예외처리 추가

dye0p committed:
feat: 24시간 형식으로 입력하지 않은 경우 예외처리 추가

dye0p committed:
feat: 존재하지 않는 크루를 찾는 경우 예외처리 추가

dye0p committed:
style: 테스트 BDD 주석 수정

dye0p committed:
test: 공휴일 출석확인 예외 테스트 결과 메세지 수정

dye0p committed:
feat: 입력 날짜 범위를 넘긴경우 예외처리 추가

dye0p committed:
docs: 등교 시간 입력 예외 수정

dye0p committed:
docs: 체크리스트 작성

dye0p committed:
refactor: Calender findBy() 메서드 반환타입 변경

dye0p committed:
refactor: 출석 상태 계산 로직의 else문 제거

- 기존의 else문을 사용하여 가독성이 저하되는 문제를
early return 하도록 변경
- 비슷한 형태의 메서드로 가독성이 저하되는 문제를 메서를 추출하도록 리팩터링

sinb57 commented on src/main/java/controller/MainController.java:
```diff
@@ -0,0 +1,108 @@
+        String feature;
+        do {
+            feature = InputView.inputFeature(todayMonth, todayDay, todayDayOfWeek);
+            if (feature.equals("1")) {
+                attendanceCheck();
+            }
+            if (feature.equals("2")) {
+                attendanceUpdate();
+            }
+            if (feature.equals("3")) {
+                attendanceRecord();
+            }
+            if (feature.equals("4")) {
+                readAbsence();
+            }
```
```suggestion
            if (feature.equals("1")) {
                attendanceCheck();
            } else if (feature.equals("2")) {
                attendanceUpdate();
            } else if (feature.equals("3")) {
                attendanceRecord();
            } else if (feature.equals("4")) {
                readAbsence();
            }
```
지금의 코드는 if-else와 다른 바 없고, 오히려 성능적으로 if-else 보다 좋지 않아요.
왜 우테코에서 else 예약어 사용에 제약을 걸었을까요? if-else를 사용하지 못하면 해당 분기 처리는 구현이 불가능할까요? 객체지향의 다형성에서 정답을 찾아보시면 좋겠어요~

dye0p replied:
단순히 else 문을 사용하지 않으면 된다고 생각했는데 결론적으로 if-else와 다른게 없네요! 피드백 반영해서 리팩터링 해보겠습니다!

sinb57 commented on src/test/java/util/FileManagerTest.java:
```diff
@@ -0,0 +1,21 @@
+
+class FileManagerTest {
+
+    @DisplayName("csv파일을 가져올 수 있다.")
+    @Test
+    void getReadFile() {
+        //given
+        Attendance attendance = FileManager.readFile();
+
+        //when & then
+        assertThat(attendance).isNotNull();
+    }
```
csv 파일을 읽는 행위에 대한 검증일까요?
반환 값을 보니 파일에 저장된 Attendance 객체를 불러오는 행위로 보이는데,
파일에 저장된 값이 객체 내부에 잘 저장되었는지 확인해야 하지 않을까요?

dye0p replied:
맞습니다! 어떻게 하면 값을 검증할 수 있을지 고민해보겠습니다!

sinb57 commented on src/main/java/controller/MainController.java:
```diff
@@ -0,0 +1,108 @@
+import view.InputView;
+import view.OutputView;
+
+public class MainController {
```
Controller는 어떠한 역할과 책임을 지녔나요? 왜 Controller라는 이름을 붙이셨나요??

dye0p replied:
사용자의 입력을 받고, 필요한 작업을 처리한 후 데이터를 받아서 결과를 보여줄 수 있도록 view와 domain을 연결해주는 징검다리 역할을 한다고 생각합니다! 그래서 제어(Control)를 의미하는 Controller라고 이름을 붙였습니다!

sinb57 replied:
답변 감사합니다! 그렇다면 Controller는 단순히 징검다리의 역할을 해야 하네요. 그런데 MainController를 보면 메서드를 많이 지니고 있어요. 징검다리 이상의, 비즈니스 로직을 수행하고 있어요. MainController를 가볍게 해보세요.

sinb57 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,92 @@
+import java.util.regex.Pattern;
+import util.Console;
+
+public class InputView {
```
InputView의 메서드들이 static으로 선언되어 있네요. 왜 static으로 선언하셨나요??

dye0p replied:
입력을 담당하는 InputView는 상태를 가지지 않기 때문에 객체간의 공유가 필요하지 않겠다고 생각해  인스턴스를 만들어 주지 않고 바로 사용할 수 있도록 static으로 선언하였습니다! 하지만 static 으로 선언하면 후에 다양한 입력을 받을 수 있는 상황에서 다형성을 활용하기 힘들다고 생각이 듭니다! view로직에서 static 메서드를 사용하는 것에 이프의 생각은 어떠신가요?!

sinb57 replied:
> 객체간의 공유가 필요하지 않겠다고 생각해 인스턴스를 만들어 주지 않고 바로 사용할 수 있도록 static으로 선언하였습니다!

그렇다면 MainController의 메서드는 왜 static으로 선언하지 않았나요?

> static 으로 선언하면 후에 다양한 입력을 받을 수 있는 상황에서 다형성을 활용하기 힘들다고 생각이 듭니다!

동의해요~!! 말씀 주신 것처럼 다형성을 활용할 수 없다는 측면에서 static은 단점이에요.
변경에 취약하기 때문에 static 메서드는 view가 아니더라도 되도록 지양하려고 해요.

dye0p replied:
> 그렇다면 MainController의 메서드는 왜 static으로 선언하지 않았나요?

그렇네요! MainController의 메서드도 객체간의 공유를 하지 않는데 굳이 view 로직까지 static으로 해야할 특별한 이유가 없다고 생각합니다! static은 객체지향에 반대되는 개념이므로 최대한 지양하는것이 좋을것 같아요!

sinb57 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,92 @@
+        }
+    }
+
+    private static boolean isCorrectFormat(final String input) {
+        String datePattern = "2[0-3]|[01][0-9]:[0-5][0-9]";
+        Pattern correctPattern = Pattern.compile(datePattern);
+
+        return correctPattern.matcher(input).find();
+    }
```
메서드를 호출할 때마다 Pattern 객체가 생성될 텐데요.
"이펙티브 자바"의 "아이템 6. 불필요한 객체 생성을 피하라" 내용을 읽어 보길 바래요~

sinb57 commented on src/main/java/domain/AbsenceHistory.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    private int calculateAbsence(AttendanceState state) {
+        return (int) attendanceResultDtos.stream().filter(dto -> dto.attendanceState().equals(state.getDescription()))
```
`Attendance#getCrewByName`와 달리 한줄로 표현하셨네요. 기준이 궁금해요~

dye0p replied:
엇! 이 부분은 인텔리제이 자동정렬을 하는 과정에서 저렇게 정렬이 된 것 같습니다ㅠㅠ 꼼꼼히 확인 해야겠네요!👍

sinb57 commented on src/main/java/domain/AbsenceHistory.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    private int calculateAbsence(AttendanceState state) {
+        return (int) attendanceResultDtos.stream().filter(dto -> dto.attendanceState().equals(state.getDescription()))
```
filter 안의 비교 로직을 보니 항상 실패할 것으로 보이는데, 확인 부탁드려요

dye0p replied:
헉 확인하지 못했네요😭
```suggestion
return (int) attendanceResultDtos.stream()
                .filter(dto ->
                        dto.attendanceState().getDescription().equals(state.getDescription()))
```
이렇게 고칠 수 있을 것 같습니다!

sinb57 replied:
아래와 같이 변경해주셨네요ㅎ 문제는 해소되었으나, 값을 꺼내어 비교하고 있네요. getter의 사용을 자제하는 규칙에 위배하고 있어요.
```java
    private int calculateAbsence(AttendanceState state) {
        return (int) attendanceResultDtos.stream()
                .filter(dto ->
                        dto.attendanceState().equals(state))
                .count();
    }
```

dye0p replied:
getter를 사용하지 않기 위해서 dto 에서 검증 로직을 넣어줘야 된다고 생각합니다! 그렇다면 데이터 전송만을 위한 dto 객체에게 값 검증 로직과 같은 로직들을 넣어줘도 괜찮을지 궁금합니다!

sinb57 replied:
말씀 주셨듯 DTO에는 로직을 만들지 않아야 해요. 도기는 이 미션에서 DTO를 왜 사용하셨나요? 지금의 AttendanceResultDto은 어디의 변경에 가장 영향을 받나요? 그렇다면 그 변경은 어디까지 전파되나요?

dye0p replied:
AttendanceResultDto를 만든 가장 큰 이유는 원하는 데이터를 저장하는 객체를 만들고 싶었기 때문입니다. 지금 사용하는 DTO의 의존성을 보면 Attendance 객체에서 가장 많이 의존하고 있습니다. 만약 DTO에서 변경이 일어난다면 Attendance 또한 수정을 해야한다고 생각해요.

sinb57 commented on src/main/java/dto/AbsenceResultDto.java:
```diff
@@ -0,0 +1,4 @@
+package dto;
+
+public record AbsenceResultDto(int attendance, int lateness, int absence, String status) {
+}
```
DTO는 무엇인가요? 어떠한 목적으로 사용하고 있으신가요??

dye0p replied:
DTO는 데이터만을 가지고 전달해주는 객체로 알고 있습니다! 해당 DTO는 제적 위험자 목록을 담을 용도로 DTO를 사용했습니다!

sinb57 replied:
DTO는 답변 주신대로 데이터 전달을 위한 객체예요. 목적에 맞게 사용은 하고 있지만, DTO라는 수단에 의해 다른 방향을 놓치고 있는 듯 하여 질문드렸어요~

AbsenceResultDto를 AbsenceHistory에서만 만들고 있어요. 그리고 AbsenceHistory는 AbsenceResultDto를 만들기 위한 객체일 뿐이네요. 그렇다면 2개의 객체로 나뉘어야 할까요?? DTO가 필요한가요??

dye0p replied:
말씀하신대로 현재 로직은 두가지 문제가 보이는것 같아요.
- 필드로 선언한 `List<AttendanceResultDto>`가 필드로써 사용되지 못하고 있음
- AbsenceResultDto를 AbsenceHistory에서만 만들고 있기 때문에 두개로 분리될 필요가 없음

그래서 결국엔 두개의 객체로 분리되지 않고 하나의 객체로 사용이 가능할 것 같다고 생각이 들어요! 우선 이렇게 정리해봤는데 이프의 생각중 다른 부분이 있다면 궁금합니다!

sinb57 replied:
저는 하나로 합치는 게 좋다고 생각해요~ 다만, 도기가 이 미션에서 생각하는 DTO가 무엇인지 궁금하네요. 아래 코멘트에서 얘기나눠보아요~ㅎ
https://github.com/woowacourse/java-attendance/pull/71#discussion_r1969875247

sinb57 commented on src/main/java/util/Console.java:
```diff
@@ -0,0 +1,28 @@
+        return getInstance().nextLine();
+    }
+
+    public static void close() {
+        if (scanner != null) {
+            scanner.close();
+            scanner = null;
+        }
+    }
```
사용하지 않는 메서드는 삭제 부탁드려요~

sinb57 commented on src/main/java/util/Console.java:
```diff
@@ -0,0 +1,28 @@
+        }
+    }
+
+    private static Scanner getInstance() {
+        if (scanner == null) {
+            scanner = new Scanner(System.in);
+        }
+        return scanner;
+    }
```
scanner 생성을 해당 메서드를 통해 처리한 이유가 궁금해요~

dye0p replied:
프리코스를 할때 우테코에서 제공한 입력 라이브러리를 비슷하게 사용하고 싶었습니다! 
그냥 Scanner를 사용할 수도 있었지만 저렇게 한다면 InputView에서 간편하게 사용할 수 있겠다고 생각했어요!

sinb57 replied:
scanner의 생성 과정을 getInstance()로 분리한 이유가 궁금했어요~
Console의 생성자에서 Scanner를 할당해도 되고 인스턴스 필드에 바로 할당해도 되는데,
왜 getInstance() 메서드를 통해 할당되도록 했는지 궁금해요

dye0p replied:
불필요한 Scanner 생성을 하고싶지 않아서 싱글턴으로 관리하고 싶어서 저렇게 구현했던것 같습니다. 
아직 싱글턴에 대해서 자세하게 이해하지 못한채로 구현해서 납득할만한 근거가 부족한것 같습니다..

다른 방법이 있다면 생성자 초기화 + static 키워드를 사용해서 상수로 관리할 수도 있을거 같은데 어떻게 생각하시나요??
두개의 방식의 차이점이 있다면 어떻게 생각하시나요?

sinb57 replied:
Console을 싱글톤으로 만들 것이 아니라면 생성자 초기화가 좋다고 생각해요. static으로 상수처리하게 되면 여러 Console 객체가 하나의 Scanner를 공유하게 될테고, 입력값이 다른 곳에서 read되어 사라지는 문제가 발생해요.

sinb57 commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,135 @@
+        Calender.validateHolyDay(date);
+
+        List<LocalDateTime> localDateTimes = attendanceMap.get(crew);
+        int i;
```
`i` 변수를 for 문과 더불어 75줄의 `localDateTimes.set(i, todayLocalDateTime);`에서도 사용하네요. 어떠한 목적의 변수인지 이름으로 파악이 어려워요. 여러 목적이 혼합된 변수는 책임을 분리하여 의도를 명시해주세요~

sinb57 replied:
attendanceRecordIndex 라고 해주셨네요! 의미 부여 좋아요~ 다만 해당 변수가 여러 행위에 관여하고 있네요. 단일 책임 원칙을 지키면 어떨까요?

dye0p replied:
메서드 분리를 통해서 해결해보겠습니다!

sinb57 commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,135 @@
+        return beforeLocalDateTime;
+    }
+
+    public List<AttendanceResultDto> readRecord(final Crew crew, int todayDay) {
+        List<LocalDateTime> localDateTimes = attendanceMap.get(crew); //해당 크루의 출석 기록
+        localDateTimes.sort(Comparator.comparing((LocalDateTime::getDayOfMonth)));
+
+        List<AttendanceResultDto> attendanceResultDtos = new ArrayList<>();
+
+        int idx = 0;
+        for (int dayIndex = 1; dayIndex < todayDay; dayIndex++) {
+            if (Calender.findBy(dayIndex).getDescription().equals("공휴일")) {
+                continue;
+            }
+
+            LocalDateTime localDateTime = null;
+            try {
+                localDateTime = localDateTimes.get(idx);
+            } catch (IndexOutOfBoundsException e) {
+                checkAbsence(dayIndex, attendanceResultDtos);
+                continue;
+            }
+
```
> 함수(또는 메서드)의 길이가 10라인을 넘어가지 않도록 구현한다.
> 함수(또는 메서드)가 한 가지 일만 하도록 최대한 작게 만들어라.

프로그래밍 요구사항에 맞춰 메서드를 수정해주세요~
https://techcourse.woowahan.com/s/W66QswFx/ls/ZXYAoQQr

sinb57 commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,135 @@
+    public Crew getCrewByName(String name) {
+        return attendanceMap.keySet()
+                .stream()
+                .filter(crew -> crew.getName().equals(name))
```
getter로 값을 꺼내어 비교하지 말고, Crew 객체에 위임하는 것은 어떤가요

dye0p replied:
Crew 객체에게 메세지를 던져 name을 비교하도록 하면 좋겠네요! 반영하겠습니다 : )

sinb57 commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,135 @@
+
+public class Attendance {
+
+    private final Map<Crew, List<LocalDateTime>> attendanceMap;
```
변수명이 자료구조에 의존하면 좋지않아요. 수정 부탁드려요~
<img width="430" alt="image" src="https://github.com/user-attachments/assets/32bb5536-e923-40ec-8324-b41ff02d0726" />

sinb57 commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    public void save(final Crew crew, final String schoolStartTime, final int todayDay) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```
재사용이 가능하면 좋겠어요~

dye0p replied:
상수로 분리하여 선언하면 좋을것 같네요! 반영하겠습니다 : )

sinb57 commented on src/main/java/domain/AttendanceTime.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public static void validateCampusTime(final LocalTime localTime) {
+        if (localTime.isBefore(LocalTime.of(8, 0)) || localTime.isAfter(LocalTime.of(23, 0))) {
```
매직넘버가 보이네요

sinb57 commented on src/main/java/domain/AbsencePolicy.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    public static String getAbsencePolicy(int absentCount, int lateCount) { //지각, 결석
+        int totalAbsentCount = absentCount + (lateCount / 3);
```
매직넘버가 보이네요

sinb57 commented on src/main/java/domain/AbsencePolicy.java:
```diff
@@ -0,0 +1,33 @@
+        this.description = description;
+    }
+
+    public int getValue() {
+        return value;
+    }
```
`WARNING.getValue()`에서 `value` 네이밍이 추상적이어서 의미를 파악하기 어렵네요. 내부 구현의 입장이 아니라, 메서드를 사용하는 외부의 시점에서 바라봐주세요~

dye0p replied:
맞습니다! 네이밍에 대해서 직관적으로 읽히도록 변경하겠습니다!👍

sinb57 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,92 @@
+    public static String inputUpdateDate() {
+        System.out.println("수정하려는 날짜(일)를 입력해 주세요.");
+        String input = Console.readLine();
+        validateDateSize(input);
+        validateInteger(input);
```
validateDateSize 에서 int으로 변경하여 검증을 진행하는데, validateInteger 에서도 동일하게 int 형에 대한 검증을 진행하네요. 어째서 검증을 분리하여 진행하는지 궁금합니다.

dye0p replied:
숫자 입력에 대한 검증과 날짜 범위에 대한 검증을 각각 해줘야겠다고 생각했습니다! 다시보니 동일한 검증을 하고있네요! 중복을 없앨 수 있도록 수정하겠습니다!

sinb57 commented on src/main/java/controller/MainController.java:
```diff
@@ -0,0 +1,108 @@
+
+    private void attendanceUpdate() {
+        String nickname = InputView.inputUpdateNickName();
+        int date = Integer.parseInt(InputView.inputUpdateDate());
```
InputView에서 무조건 String 값을 넘겨받아야 할까요?

dye0p replied:
InputView의 검증 로직에서 이미 파싱을 진행해주고 있네요! 그럼 검증로직에서 파싱된 값을 validate로직에서 리턴해줘도 괜찮을지 궁금합니다!

sinb57 replied:
InputView에서 타입에 대한 검증과 더불어 해당 타입을 반환해도 되지 않을까 합니다. 꼭 String 값을 반환받아야 할까요?
예를 들어 `InputView#inputUpdateDate` 메서드가 LocalDateTime을 반환해도 되지 않을까요?

dye0p replied:
이 부분에 대해서 고민이 있었는데 피드백 감사합니다! 말씀하신대로 Controller의 입장에서는 시간을 받을때 String 보단 LocalDateTime 타입으로 받아도 될것 같아요! 이미 InputView에서 검증 로직을 수행하고 반환하는데 굳이 Controller에서 한번 더 검증과 파싱을 하는건 불필요하다고 생각이 들어요! 좋은 피드백 감사합니다 : )

sinb57 commented on README.md:
```diff
@@ -1,3 +1,31 @@
+- [x] 특정 크루의 출석 기록 출력
+- [x] 제적 위험자 결과 출력
+
+## 예외 처리
+
+- 입력 형식이 맞지 않는 경우, IllegalArgumentException을 발생시킨 후 애플리케이션을 종료시킨다.
```
<img width="833" alt="image" src="https://github.com/user-attachments/assets/cfe051aa-d1ea-4102-b822-c6269160c1bf" />
<img width="803" alt="image" src="https://github.com/user-attachments/assets/350d60f3-19a2-4ad3-98f7-0be60c32be41" />

README에 예외 케이스를 정리해주면 좋겠어요~ 더불어 예외 케이스에 대한 단위 테스트도 작성해주세요.

dye0p replied:
요구사항 정리와 단위테스트가 부족했네요! 피드백 반영하여 수정하겠습니다!

sinb57 reviewed (changes requested):
안녕하세요 도기~ 미션 진행하시느라 고생 많으셨어요! 아래의 피드백과 더불어 코멘트 확인하여 반영해주세요.

- 예외 케이스
- if 분기를 개선해주세요.
- 메서드의 길이가 10라인을 넘어가지 않도록 해주세요.
- 객체를 재사용하여, 불필요한 객체 재생성을 줄여보세요.
- 매직넘버가 여럿 보이네요. 의미를 부여하여 개선해보세요.
- 열려있는 모든 메서드는 단위테스트를 작성해주세요.
- getter의 사용을 자제해주세요.

---

질문이 많지만 내용이 같아보여서 답변을 묶어서 드립니다.
그리고 도기의 생각을 먼저 알아야 적절한 답변이 가능한 부분이 있다보니, 반문을 드리기도 했습니다

> - 가장 컷던 고민은 테스트를 작성할 때 어떤 크기의 테스트 부터 작성할지 고민이 되었습니다. 큰 단위의 테스트 부터 작성한다면 고려해야할 사항이 많아진다는 단점이 있었고, 작은 단위의 테스트 부터 작성한다면 이것들을 나중에 어떻게 조립할지 고민이었습니다. 이 부분에 대한 이프의 생각은 어떤지 궁금합니다!
> - 어떤 객체에 어떤 테스트를 만들어야 하지? TDD 특성상 먼저 테스트를 만들어야 하는데, 이때 해당 테스트는 어떤 객체가 가지는 책임이지? 라는 생각이 들었습니다. 그래서 처음엔 예상하여 테스트를 작성했지만, 점점 시간이 지나고 크기가 커지다보니 원래 작성했던 테스트가 완전히 필요없어져 버리는 경우가 빈번하게 일어났습니다. 코드를 작성할수록 필요한 객체가 많아지게 되고, 필요한 테스트를 하기위한 객체를 판단하기가 어려워졌습니다. 이 부분에 대한 이프의 생각은 어떤지 궁금합니다!

구현에 앞서 대략적인 설계를 먼저 구상 하고서 접근해야 수월하지 않을까 싶어요. 요구사항을 구현하기 위해서는 어떠한 정보와 행위들이 필요하고 이를 위해서는 어떠한 객체들을 선언하여 유기적인 결합 구조를 구성할 지 등 각 객체의 명확한 역할과 책임을 부여한 후에야 제대로 된 TDD가 가능하다고 생각해요. 그렇지 않으면 객체의 역할과 책임이 모호해지면서 계속된 수정으로 진행이 안될거에요.

> TDD 적용의 어려움으로 TDD의 장점을 제대로 느끼지 못했습니다. 아직은 TDD가 어떻게 객체지향적인 코드를 작성하는 것인지 제대로 와닫지가 않았습니다. TDD를 잘 하기위한 이프만의 팁이 있다면 알려주세요!🙇‍♂️

도기는 테스트코드의 장점을 무엇이라 생각하시나요?

> - TDD는 핵심 도메인 로직의 테스트를 먼저 작성하면서 개발을 하는 것 이라고 알고 있습니다. 하지만 점점 도메인의 규모가 커지고 테스트가 복잡해짐으로 인해서 이러한 테스트가 이후 뷰(view)와 컨트롤러(Controller)에 잘 조합이 될지가 불안요소로 다가왔습니다. 도메인 테스트만 하다보니 콘솔에 직접 입력하여 눈으로 실행결과를 확인하고 싶다는 생각이 많이 들었습니다. 이러한 고민에 대한 이프의 생각은 어떤지 궁금합니다!
> view에서 validation 로직을 작성해줘도 괜찮을까?

로또 미션에서 학습한 내용에 대해 우선 질문드릴게요. 왜 비즈니스 로직과 UI 로직을 분리해야 할까요? 어떠한 문제가 있기 때문에 로직의 분리를 추천하는 걸까요?

dye0p commented:
1단계 리뷰를 해주셔서 감사합니다 이프!🙇‍♂️ 고칠부분이 많은것 같아요! 말씀해주신 피드백 반영해서 다시 리뷰요청 하겠습니다! 수고하셨습니다👍

dye0p committed:
fix: calculateAbsence() 메서드 스트림 filter 로직 수정

dye0p committed:
refactor: Console 객체 사용하지 않는 메서드 제거

dye0p committed:
refactor: for문 인덱스 변수명 명확하게 변경

dye0p committed:
refactor: Crew 객체가 name을 비교하도록 책임 변경

dye0p committed:
refactor: 변수명에 자료구조명 제거

- 변수가 자료구조에 종속되는 문제가 발생해 변수명을 수정하였음

dye0p committed:
refactor: LocalDateTime 포멧 로직 상수화 및 메서드 분리

dye0p committed:
fix: 출석 기록 조회 포맷 수정

- 크루별 출석 기록 조회시 출력 형태 오류를 수정함

dye0p committed:
refactor: 매직넘버 & 매직스트링 상수화

dye0p committed:
refactor: 숫자 검증 로직 중복메서드 제거

- 동일한 기능을 하는 중복된 메서드를 하나의 메서드로 리팩터링
- 날짜 매직넘버 상수화

dye0p committed:
refactor: 날짜 입력 반환타입 int형으로 변경

- 입력단에서 파싱 검증을 하고, 컨트롤러에서 한번 더 파싱을 하여 불필요한 중복이 있던 로직을 파싱 검증 후 반환하여 컨트롤러에서 파싱을 하지 않고 불필요한 중복을 줄였음

dye0p committed:
docs: 기능 요구사항 목록에 예외상황 목록 추가

dye0p committed:
test: 테스트 DisplayName 수정

dye0p committed:
fix: 제적 목록을 가져오는 로직에서 오늘 날짜로 인식하지 못한 문제 수정

- 파라미터로 임의의 숫자를 넣어주는 것을 오늘 날짜를 가지는 변수로 변경

dye0p committed:
refactor: FileManage 객체의 파일 경로를 외부에서 주입받도록 변경

dye0p committed:
refactor: readFile() 메서드 예외타입 변경(RunTime -> IllegalArgument)Exception

dye0p committed:
refactor: AttendanceTest 테스트 픽스처 추가

dye0p committed:
refactor: AbsencePolicy 지각횟수 변수명 명확하게 변경

dye0p committed:
refactor: getAbsencePolicy() 메서드 반환타입 변경(String -> AbsencePolicy)

dye0p committed:
refactor: 24시 형식 패턴 객체 상수로 선언

- 생성비용이 높은 Pattern 객체를 매번 생성해는 문제를
상수로 선언하여 재사용이 가능하도록 리팩터링 하였음

dye0p committed:
test: FileManage 파일 읽어오기 테스트 코드 검증 결과 추가

dye0p committed:
refactor: 입력한 기능에 따른 분기처리를 제거

- 입력한 기능을 처리하는 부적절한 분기문이 많아,
입력한 기능에 따른 분기 처리를 인터페이스를 사용하여 처리하도록 변경하였음

dye0p committed:
feat: 기능 타입 FeatureType Enum 추가

- 문자열로 구분되어서 변경에 취약했던 문제를 기능 타입을 가지는 Enum 객체를 생성하여 해결
- 입력 타입 검증 로직을 뷰(View)에서 FeatureType 으로 책임 이동

dye0p committed:
refactor: 출력 형식 알맞게 리팩터링

dye0p committed:
feat: 프로그램 종료시 안내문 출력 기능 추가

dye0p committed:
docs: 파일 입출력 예외 완료사항 반영

dye0p committed:
refactor: today 포멧팅 문자열 상수화

dye0p committed:
refactor: 날짜의 공휴일 판단 로직을 Calender 객체의 책임으로 이동

dye0p committed:
refactor: IndexOutOfBoundsException을 사용하지 않고 리스트 크기를 비교하는 방식으로 리팩터링

- 불필요한 예외 발생을 없애고 리스트 크기를 비교하는 방식으로 변경

dye0p committed:
refactor: Attendance 메서드 분리 리팩터링

dye0p committed:
refactor: 메서드 분리 및 위치 정렬 리팩터링

dye0p committed:
refactor: Command 인터페이스 패키지 이동(controller -> domain)

dye0p committed:
refactor: featureMap 변수의 자료구조명 제거

dye0p committed:
refactor: 매직넘버 상수화

dye0p committed:
refactor: 값 비교 로직 객체 비교로 리팩터링

dye0p commented:
안녕하세요 이프! 피드백 반영이 늦어져서 죄송합니다 😢 피드백 주신 부분들을  반영해보았습니다! 리팩터링 할 사항이 많아, 모두다 하지는 못하고 우선 다시 리뷰요청 드립니다! 혹시 피드백이 필요한 부분이 있다면 말씀해주세요! 그리고 피드백 중  질문에 대한 답변과 피드백 내용은 아니지만 궁금했던 사항을 말씀드릴게요!

> 도기는 테스트코드의 장점을 무엇이라 생각하시나요?

테스트 코드의 장점은 많지만  그 중에서도 안정감과 테스트시 발생하는 비용을 줄여주는것이 가장 크다고 생각합니다! 성공하는 테스트 코드를 바탕으로 리팩터링을 수월하게 할 수 있다고 생각해요. 그리고 수동테스트를 하지않고 빠르게 테스트를 할 수 있다는 장점도 있다고 생각합니다. 수동 테스트를 하게되면 실수를 하게될수도 있고 많은 예외상황을 고려하지 못할 것이라고 생각합니다. 

> 로또 미션에서 학습한 내용에 대해 우선 질문드릴게요. 왜 비즈니스 로직과 UI 로직을 분리해야 할까요? 어떠한 문제가 있기 때문에 로직의 분리를 추천하는 걸까요?

변경에 취약하다는 것이 가장 큰 이유라고 생각합니다. UI로직은 언제든 변경될 수 있습니다. 만약 변경에 취약한 UI로직과 비즈니스 로직이 결합되어있다면 UI로직의 변경이 일어날 때마다 비즈니스 로직도 함께 변경된다고 생각합니다. 그렇다면 UI로직에서 검증후에 파싱을 하여 필요한 값을 리턴해주는것은 이프는 어떻게 생각하나요?

## 🤔 고민 포인트
### 도메인 객체의 DTO 의존

저번 로또 미션 수행중 도메인 객체는 DTO에 의존하면 안된다. (도메인 객체의 import에 DTO가 있으면 안된다) 라고 피드백을 받았습니다.
별도의 Service 로직이 없는 현재의 코드에서 DTO 생성하거나 매핑하는 작업은 도메인 객체가 아니라면 어디서 해줘야할지 고민이 되었습니다.
이 부분에 대한 이프의 생각이 궁금합니다!

sinb57 commented on src/test/java/util/FileManagerTest.java:
```diff
@@ -0,0 +1,43 @@
+        String filePath = "src/main/resources/attendances.csv";
+
+        //when
+        Attendance attendance = FileManager.readFile(filePath);
```
아래의 경우 테스트코드를 작성하면 좋을까요?
- 파일 안의 데이터를 읽어와 객체를 잘 생성하는가(객체 내부의 값과 비교)
- 파일 안의 데이터 포맷이 잘못된 경우 에러가 발생하는가

dye0p replied:
정상 동작의 경우와 파일 포맷 과정중 예외가 발생할 수도 있겠네요! 반영해보겠습니다!

sinb57 commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,145 @@
+        attendances.put(crew, localDateTimes);
+    }
+
+    public LocalDateTime update(final Crew crew, final String updateTime, final int date) {
```
두번째 인자 updateTime이 String 타입이네요. 왜 String 타입인가요? 해당 메서드를 사용하는 외부의 입장에서는 String으로 값을 넘겨주는 게 자연스러울까요, LocalDateTime으로 값을 전달하는게 자연스러울 까요? 지금은 오히려 String 타입이기 때문에, 잘못된 포맷으로 입력된 경우에 대한 에러 테스트코드가 필요합니다.

dye0p replied:
값을 전달받는 메서드의 관점에서 생각해보면 시간을 받을 때 말씀하신 것처럼 String 타입보단 LocalDateTime을 받는것이 더 자연스러울 것 같네요! 메서드에서는 어떤 값이 들어올지 정확히 알 수 없지만 파라미터 타입으로 예상할 수 있어야 된다고 생각합니다! 피드백 반영하여 수정하겠습니다 👍

sinb57 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,85 @@
+                beforeDateTime.getDayOfMonth());
+
+        AttendanceState afterAttendanceState = AttendanceState.findStateBy(afterDateTime.toLocalTime(),
+                afterDateTime.getDayOfMonth());
```
LocalDateTime을 인자로 받도록 `AttendanceState#findStateBy` 메서드를 오버로딩하면 어떨까요?

dye0p replied:
평소 오버로딩을 잘 사용하지 않아서 떠올리지 못했는데 좋은 방법인것 같아요! 메서드 오버로딩을 적용하면서 느낀 장단점에 대한 이프의 생각도 궁금합니다! 

장점: 메서드의 유연성이 증가한다.
단점: 동일한 네이밍의 메서드로 인한 혼란

sinb57 replied:
말씀주신 단점의 경우, 메서드 체이닝을 통해 해소할 수 있어 보여요~!!

sinb57 commented on src/main/java/domain/AbsencePolicy.java:
```diff
@@ -0,0 +1,38 @@
+        this.description = description;
+    }
+
+    public static AbsencePolicy getAbsencePolicy(int absentCount, int lateCount) {
+        int totalAbsentCount = calculateTotalAbsentCount(absentCount, lateCount);
+
+        return Stream.of(DISMISSED, INTERVIEW, WARNING)
+                .filter(policy -> totalAbsentCount >= policy.lateCount)
+                .findFirst()
+                .orElse(PASS);
+    }
+
+    private static int calculateTotalAbsentCount(final int absentCount, final int lateCount) {
+        return absentCount + (lateCount / ABSENCE_CONVERSION_RATE);
+    }
```
`calculateTotalAbsentCount(final int absentCount, final int lateCount)` 메서드를 분리한 이유가 있나요?

dye0p replied:
총 결석 횟수를 계산하는 로직을 하나의 행동으로 봤습니다! 나중에 총 결석 횟수 계산 방법이 변경될 것을 고려하여 하나의 메서드로 분리 했습니다!

sinb57 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,41 @@
+
+    public Crew(final String name) {
+        this.name = name;
+    }
```
내부에서만 사용한다면 되도록 닫아주세요~

sinb57 commented on src/main/java/controller/MainController.java:
```diff
@@ -0,0 +1,119 @@
+            feature = InputView.inputFeature(todayMonth, todayDay, todayDayOfWeek);
+            executeFeature(feature);
+        } while (isExit(feature));
+    }
```
여기서 호출하는 executeFeature 메서드를 살펴보기 위해서는 비교적 먼 거리를 이동해야 하네요. 크진 않지만 시간과 노력이 소모되다 보니, 수직거리를 고려해주시면 좋아요!
시간 괜찮을 때 "클린코드(p.101)"를 읽어보세요~

sinb57 commented on src/main/java/controller/MainController.java:
```diff
@@ -0,0 +1,119 @@
+    private String todayDayOfWeek;
+    private Attendance attendance;
+
+    private final Map<FeatureType, Command> features = Map.of(
+            FeatureType.ATTENDANCE_CHECK, this::attendanceCheck,
+            FeatureType.ATTENDANCE_UPDATE, this::attendanceUpdate,
+            FeatureType.ATTENDANCE_RECORD, this::attendanceRecord,
+            FeatureType.READ_ABSENCE, this::readAbsence
+    );
```
FeatureType 이라는 Enum을 만드셨는데, 해당 Enum을 활용하면 해당 Map은 불필요하지 않을까요?

dye0p replied:
해당 피드백을 보고 지워보려고 했지만 방법을 찾지 못했습니다..! Map을 사용하지 않고 Enum만 사용하게 되면 Controller의 메서드가 Enum에 의존되는 상황이 발생했습니다. 도메인 객체는 Controller와 view 등에 의존하면 안된다고 알고 있는데 이프는 어떻게 생각하시나요?

sinb57 replied:
음 명령어가 "1" -> "a" , "2" -> "b"와 같이 변경된다면 FeatureType에 변경은 없을까요? 이는 view의 변경이 도메인에 변경을 가하는 상황 아닐까요? FeatureType는 도메인일까요?

dye0p replied:
말씀하신 것처럼 명령어가 변경되면 FeatureType도 함께 변경되어야 한다고 생각합니다. FeatureType의 로직을 보면 단순하게 들어온 입력값에 대한 매핑만 이루어지고 있기에 도메인 객체라고 보기가 힘들 것 같습니다. 

도메인 객체가 되기 위한 요건은  `상태의 변경이 이루어지는 비즈니스 로직이 존재하는가?`라고 생각합니다.

따라서 단순히 입력값의 매핑만을 한다면, 입력값에 따라서 Controller의 메서드를 호출할 수 있다고 생각합니다!

sinb57 commented on src/main/java/util/FileManager.java:
```diff
@@ -0,0 +1,65 @@
+        }
+
+        return new Attendance(attendances);
+    }
```
Map<Crew, List<LocalDateTime>> 자료구조를 만들어 생성한 후 Attendance 객체에 넣은 것이 아니라, Attendance 객체를 만들고 값을 하나씩 넣어가면 어떨까요? 지금의 코드가 잘못된 것은 아니나 객체지향의 코드를 연습하면 좋겠습니다.

dye0p replied:
맞습니다! 지금껏 외부에서 주입에 필요한 컬렉션을 생성하여 주입해주었습니다. 객체의 입장에서 생각해보니 Map의 상태를 알아야 하는것은 Attendance 객체이므로 객체가 스스로 Map을 생성하도록 한다면 캡슐화가 지켜질 것이라고 생각합니다!

sinb57 commented on src/main/java/domain/Calender.java:
```diff
@@ -0,0 +1,44 @@
+    WED("수요일", List.of(4, 11, 18)),
+    THU("목요일", List.of(5, 12, 19, 26)),
+    FRI("금요일", List.of(6, 13, 20, 27)),
+    HOLY("공휴일", List.of(1, 7, 8, 14, 15, 21, 22, 25, 28, 29));
```
12월의 모든 날짜를 Calender 에 담아서 관리하는 지금 방법은 자칫 휴먼 에러를 일으킬 가능성이 높아요. 12월의 모든 날짜를 담고 있다고 항상 보장할 수 있나요?

sinb57 commented on src/main/java/domain/Calender.java:
```diff
@@ -0,0 +1,44 @@
+        if (dayOfWeek.description.equals("공휴일")) {
+            throw new IllegalArgumentException("공휴일에는 출석을 할 수 없습니다.");
+        }
+    }
```
java.time.DayOfWeek를 활용해보실래요?

sinb57 commented on src/main/java/domain/Command.java:
```diff
@@ -0,0 +1,6 @@
+@FunctionalInterface
+public interface Command {
+    void execute();
+}
```
함수형 인터페이스를 사용하셨군요! 👍

sinb57 reviewed (changes requested):
안녕하세요 도기! 피드백 반영하느라 고생 많으셨어요! 추가 코멘트 확인 부탁드려요~

---

> 2024년 12월 한 달 동안 시범적으로 최소한의 기능을 갖춘 출석 시스템을 개발하여 출석을 체계적으로 관리하기로 하였습니다

요구사항을 보면, 현재는 12월 한정 시범 운영이지만 시스템이 확장될 가능성을 엿볼 수 있어요. 이 경우 현재의 구현에서 확장에 대응하려면 어떠한 변경이 있을지 생각해보시겠어요? 과연 현재의 코드는 변경에 열려 있나요?

---

>> 도기는 테스트코드의 장점을 무엇이라 생각하시나요?
>
> 테스트 코드의 장점은 많지만 그 중에서도 안정감과 테스트시 발생하는 비용을 줄여주는것이 가장 크다고 생각합니다! 성공하는 테스트 코드를 바탕으로 리팩터링을 수월하게 할 수 있다고 생각해요. 그리고 수동테스트를 하지않고 빠르게 테스트를 할 수 있다는 장점도 있다고 생각합니다. 수동 테스트를 하게되면 실수를 하게될수도 있고 많은 예외상황을 고려하지 못할 것이라고 생각합니다.

테스트 성공과 실패로부터 피드백을 받으시는군요! 테스트 코드의 피드백에는 여럿 관점이 있지만, 그 중 구현에 대한 피드백이 존재한다고 봐요. TDD의 첫 스텝으로 실패하는 테스트 코드를 작성하잖아요. 이 때는 도기가 말한 테스트의 성공과 실패가 무의미한데, 어째서 TDD는 실패하는 테스트 코드를 먼저 작성할까요?

>>로또 미션에서 학습한 내용에 대해 우선 질문드릴게요. 왜 비즈니스 로직과 UI 로직을 분리해야 할까요? 어떠한 문제가 있기 때문에 로직의 분리를 추천하는 걸까요?
>
> 변경에 취약하다는 것이 가장 큰 이유라고 생각합니다. UI로직은 언제든 변경될 수 있습니다. 만약 변경에 취약한 UI로직과 비즈니스 로직이 결합되어있다면 UI로직의 변경이 일어날 때마다 비즈니스 로직도 함께 변경된다고 생각합니다. 그렇다면 UI로직에서 검증후에 파싱을 하여 필요한 값을 리턴해주는것은 이프는 어떻게 생각하나요?

맞아요~ 비즈니스 로직과 UI 로직을 분리하는 주된 이유는 변경에 취약하기 때문이죠. UI 로직은 변경될 수 있어도 비즈니스의 핵심 로직은 변경이 없어야 하죠. 다른 얘기지만, 그렇다면 책임의 기준은 무엇이어야 할까요? 객체는 무엇을 기준으로 분리하는 게 좋을까요?

`그렇다면 UI로직에서 검증후에 파싱을 하여 필요한 값을 리턴해주는것은 이프는 어떻게 생각하나요?`
=> 무엇을 검증하지만, 파싱하여 타입을 변환하여 반환하는 것은 문제 없다고 봐요. 앞서 말씀주셨듯 비즈니스 로직과 UI 로직은 분리하여 생각해야 해요. 이때 비즈니스 로직 입장에서는 String 타입으로 받는 것이 부자연스러운 경우가 꽤 있어요. 무의식 중에 이미 UI로직의 영향을 받았을 거예요. 도기는 구현할 때 입출력에 대한 생각을 배제하고 오로지 비즈니스 로직에 대해서만 생각하여 구현해보셨나요? 도메인 객체의 메서드가 String 타입으로 열려있고, 넘겨받아 검증 및 파싱하는 것은 이미 입출력의 영향을 받았다고 생각해요.

> 도메인 객체의 DTO 의존
> 저번 로또 미션 수행중 도메인 객체는 DTO에 의존하면 안된다. (도메인 객체의 import에 DTO가 있으면 안된다) 라고 피드백을 받았습니다. 별도의 Service 로직이 없는 현재의 코드에서 DTO 생성하거나 매핑하는 작업은 도메인 객체가 아니라면 어디서 해줘야할지 고민이 되었습니다. 이 부분에 대한 이프의 생각이 궁금합니다!

질문에서 답을 구하지 못해, 반문 드립니다. 왜 도메인이 DTO에 의존하면 안되나요?

dye0p committed:
refactor: MainController 메서드 수직거리 고려 리팩터링

- 호출하는 메서드와 호출되는 메서드 간의 수직거리가 길어 가독성이 떨어지는 문제 수정

dye0p committed:
refactor: View 클래스 메서드의 static 키워드 제거

- 변경을 고려해 다형성을 저해하는 static 키워드를 제거하고, MainController 에서 주입받도록 변경

dye0p committed:
feat: 날짜를 상태로 가지는 DateProvider 객체 추가

- Controller에서 날짜를 생성하고 있어, 통합 테스트가 힘들었던 부분을 외부에서 날짜를 제공해주는 DateProvider 객체를 생성하여 해결

dye0p committed:
refactor: 오늘 날짜의 요일을 가져오는 로직을 DateProvider에서 수행하도록 변경

dye0p committed:
refactor: 시간 입력시 LocalDate 타입으로 반환하도록 변경

- 기존 입력 시간 반환타입이 String으로 어색헀던 반환 타입을 LocalDate 타입으로 포장하여 반환하도록 변경
- Attendance에서 출석, 수정 메서드의 파라미터 타입으로 받았던 String 타입은 어색하다는 문제점을 LocalDateTime 타입으로 받아서 수행하도록 변경

dye0p committed:
refactor: Attendance 출석수정 update 메서드 로직 리팩터링

- null 사용을 하지 않기 위해서 반복문을 스트림으로 변경 하였음

dye0p committed:
refactor: Crew 객체 생성자 접근 제어자 변경

- 외부에서 사용되지 않는 생성자이기 때문에 private로 변경하였음

dye0p committed:
refactor: findStateBy 메서드 오버로딩

dye0p committed:
refactor: DateProvider 메서드 리팩터링

- 기존의 오버로딩 된 createLocalDateTimeBy 메서드의 중복이 있는 부분을 별도의 메서드로 분리함

dye0p committed:
refactor: 값을 꺼내어 비교하지 않고 Dto 내부에서 값을 비교하도록 변경

- getter 사용을 자제하기 위해서 dto 내부에서 state 값을 비교함

dye0p committed:
refactor: 불필요한 AbsenceResultDto를 제거하고 AbsenceHistory를 반환하도록 변경

dye0p committed:
feat: FileManager 잘못된 구분자 예외 추가

dye0p committed:
feat: FileManager 파싱값 공백제거 추가

dye0p committed:
refactor: 출력형식 수정(공백 추가)

dye0p committed:
refactor: Attendance 객체의 필드를 직접 생성하도록 변경

- 외부에서 Attendance의 필드를 생성하여 주입시켜준다면 Attendance는 자신의 필드를 외부의 값에 의존하게 되는 문제점이 있었음
- 외부에서 Map<>을 생성하여 주입하지 않고, Map<>을 Attendance 객체 내부에서 생성하도록 변경하였음

dye0p committed:
refactor: 한개의 라인 스킵 메서드 분리

dye0p committed:
refactor: Calender Enum 필드 변경

- 기존 Calender 객체는 날짜를 직접 입력 받아서 안정적이지 않다는 단점이 있었다. 필드에 날짜를 직접 입력하지 않고, DayOfWeek 타입을 사용하여 요일을 비교하도록 변경했다.

dye0p reviewed (commented):
안녕하세요 이프! 피드백 감사합니다🙇‍♂️ 말씀해주신 피드백을 최대한 반영해보았습니다!

## 리팩터링 사항
- 12월에 한정되지 않고 변경에 열려있도록 외부에서 날짜를 주입받도록 수정
- Controller를 가볍게
- getter 사용 지양
- 불필요한 Dto 제거
- 입력 날짜, 시간을 자연스럽게 LocalDateTime으로 포장하여 반환
- Attendance 객체가 스스로 필드를 관리하도록 변경
- Calender 객체의 날짜를 DayOfWeek 타입으로 변경

## 리팩토링 하지 못한 부분

> FeatureType 이라는 Enum을 만드셨는데, 해당 Enum을 활용하면 해당 Map은 불필요하지 않을까요?

이 부분에 대해서 Enum을 활용하려면 Controller의 메서드가 Enum에 의존되는 상황이 발생해 더 좋은 방법을 찾지 못했습니다..!

## 질문에 대한 답변

> 테스트 성공과 실패로부터 피드백을 받으시는군요! 테스트 코드의 피드백에는 여럿 관점이 있지만, 그 중 구현에 대한 피드백이 존재한다고 봐요. TDD의 첫 스텝으로 실패하는 테스트 코드를 작성하잖아요. 이 때는 도기가 말한 테스트의 성공과 실패가 무의미한데, 어째서 TDD는 실패하는 테스트 코드를 먼저 작성할까요?

실패하는 코드를 먼저 작성함으로써 실패하는 경우를 인지할수 있을거라고 생각이 들었습니다!

> 맞아요~ 비즈니스 로직과 UI 로직을 분리하는 주된 이유는 변경에 취약하기 때문이죠. UI 로직은 변경될 수 있어도 비즈니스의 핵심 로직은 변경이 없어야 하죠. 다른 얘기지만, 그렇다면 책임의 기준은 무엇이어야 할까요? 객체는 무엇을 기준으로 분리하는 게 좋을까요?

객체가 외부에 제공할 수 있는 행동을 기준으로 분리하는것이 좋다고 생각합니다. 객체는 서로 필요한 요청을 주고 받으며 상호작용 하지만 어울리지 않는 행동이 객체에게 있다면 외부의 영향을 많이 받게 되고 응집도를 해친다고 생각해요!
아직 책임의 기준을 명확하게 세우는 것이 조금 힘든것 같아요. 이프는 무엇을 기준으로 객체를 분리하나요?

> 질문에서 답을 구하지 못해, 반문 드립니다. 왜 도메인이 DTO에 의존하면 안되나요?

DTO는 저수준 모듈, 도메인은 고수준 모듈 입니다. 고수준 모듈은 저수준 모듈에 의존해서는 안된다고 알고 있습니다. 만약 DTO가 변경된다면 도메인 객체까지 영향이 갈 것이라고 생각합니다. 그렇기 때문에 도메인이 DTO에 의존하면 안된다고 생각해요!

sinb57 commented on src/main/java/util/FileManager.java:
```diff
@@ -0,0 +1,67 @@
+
+    public static Attendance readFile(String filePath) {
+        try {
+            BufferedReader br = new BufferedReader(new FileReader(filePath));
```
자원은 close해주는 게 좋아요!! 이와 관련해서 try-with-resources에 대해 찾아보시는 걸 추천드려요

dye0p replied:
그렇군요! 자원 사용 후 close 해줘야 함을 알게되었습니다! try-with-resources에 대해서 찾아보니 이전엔 null검사를 해주어 수동으로 자원을 close 해줘야했던 것에 비해 자동으로 close를 해주는 것이군요! 피드백 감사합니다👍

sinb57 reviewed (changes requested):
안녕하세요 도기~ 질문이 여럿 남아있어서 답변 및 질문 남겼어요! 확인 부탁드려요~

---

>> 테스트 성공과 실패로부터 피드백을 받으시는군요! 테스트 코드의 피드백에는 여럿 관점이 있지만, 그 중 구현에 대한 피드백이 존재한다고 봐요. TDD의 첫 스텝으로 실패하는 테스트 코드를 작성하잖아요. 이 때는 도기가 말한 테스트의 성공과 실패가 무의미한데, 어째서 TDD는 실패하는 테스트 코드를 먼저 작성할까요?
>
> 실패하는 코드를 먼저 작성함으로써 실패하는 경우를 인지할수 있을거라고 생각이 들었습니다!

실패하는 경우를 인지한다는 의미는 예외케이스를 파악할 수 있다는 말일까요??

>> 맞아요~ 비즈니스 로직과 UI 로직을 분리하는 주된 이유는 변경에 취약하기 때문이죠. UI 로직은 변경될 수 있어도 비즈니스의 핵심 로직은 변경이 없어야 하죠. 다른 얘기지만, 그렇다면 책임의 기준은 무엇이어야 할까요? 객체는 무엇을 기준으로 분리하는 게 좋을까요?
>
> 객체가 외부에 제공할 수 있는 행동을 기준으로 분리하는것이 좋다고 생각합니다. 객체는 서로 필요한 요청을 주고 받으며 상호작용 하지만 어울리지 않는 행동이 객체에게 있다면 외부의 영향을 많이 받게 되고 응집도를 해친다고 생각해요!
아직 책임의 기준을 명확하게 세우는 것이 조금 힘든것 같아요. 이프는 무엇을 기준으로 객체를 분리하나요?

책임이 무엇인지 정의하면 객체에 대해 한발짝 다가갈 수 있어요.
스스로 정의할 수 있도록 질문을 드려볼게요! 제가 생각하는 기준은 나중에 알려드릴게요.
- 객체가 지니는 행동은 어떠한 기준으로 부여하고 분리하나요?
- 객체에게 어울리지 않은 행동은 어떻게 알 수 있을까요?
- 응집도는 무엇일까요?

>> 질문에서 답을 구하지 못해, 반문 드립니다. 왜 도메인이 DTO에 의존하면 안되나요?
> 
> DTO는 저수준 모듈, 도메인은 고수준 모듈 입니다. 고수준 모듈은 저수준 모듈에 의존해서는 안된다고 알고 있습니다. 만약 DTO가 변경된다면 도메인 객체까지 영향이 갈 것이라고 생각합니다. 그렇기 때문에 도메인이 DTO에 의존하면 안된다고 생각해요!

고수준과 저수준은 무엇을 의미하나요?? 고수준 모듈이 저수준 모듈에 의존하면 안되는 이유는 무엇일까요? DTO의 변경이 도메인 객체에 영향을 주면 안되나요? DTO는 view의 영역인가요??

dye0p committed:
refactor: try-with-resources를 이용하여 자원을 close 해주도록 변경

- 자원을 사용하고 close를 해주지 않게되어 메모리 낭비가 있었던 문제를 try-with-resources를 사용하여 자동으로 자원을 close 해주도록 변경

dye0p reviewed (commented):
이프! 남겨주신 피드백 부분 수정하고, 질문에 대한 답변과 함께 리뷰요청 보내겠습니다!

> 실패하는 경우를 인지한다는 의미는 예외케이스를 파악할 수 있다는 말일까요??

네 맞습니다! 실패하는 테스트 코드로 어떤 케이스에서 실패하는지 알 수 있을것 같아요! 혹시 다른 의미가 있을까요?

> 객체가 지니는 행동은 어떠한 기준으로 부여하고 분리하나요?

명확한 기준이 없어 사람마다 다르지만 외부에 제공해야 하는 기능을 수행할 가장 어울리는 객체를 선택합니다! 객체의 선택 기준은 해당 기능을 수행하기 위해서 얼마나 많은 정보를 알고있는가? 로 판단하는것 같아요! 

> 객체에게 어울리지 않은 행동은 어떻게 알 수 있을까요?

객체가 해당 행동을 수행하기에 적절한 정보나 상태가 없다면 해당 객체는 행동을 수행하기가 어색하다고 판단하는것 같습니다!

응집도는 무엇일까요?

`객체가 얼마나 자신의 역할을 잘 하고 있는가?`로 생각됩니다. 객체가 자신의 일을 외부에 알리지 않고 잘 수행한다면 응집도가 높다고 생각합니다!

> 고수준과 저수준은 무엇을 의미하나요?? 고수준 모듈이 저수준 모듈에 의존하면 안되는 이유는 무엇일까요? DTO의 변경이 도메인 객체에 영향을 주면 안되나요? DTO는 view의 영역인가요??

명확한 차이가 와닿지 않지만 고수준은 추상적인것이고 저수준은 그보다 구체화된 것 이라고 생각합니다. 구체는 변경이 일어날 가능성이 높기때문에 만약 추상이 구체에 의존하게 된다면 변경에 유연하게 대응할 수 없을거라 생각합니다. 
DTO에 변경이 일어났을때 도메인까지 변경이 일어나게 된다면 DTO가 변경될때 마다 도메인 객체를 수정해야하고, 프로그램 규모가 커지면 도메인과 DTO의 책임이 분리되지 않아 유지보수에 어려움이 생길 거라고 생각해요. 
DTO는 View의 영역보단 도메인과 View의 사이에 존재하는거라고 생각합니다. View에 도메인을 직접 넘겨서 사용하는것 보단 DTO를 한번 거쳐 필요한 데이터만 안전하게 넘길 수 있다고 생각해요!

리뷰 수고 많으셨습니다 : )

dye0p committed:
refactor: 입력한 기능 값에 따라서 기능을 수행하도록 FeatureType 리팩터링

dye0p commented:
이프! 남겨주신 피드백 부분 수정하고, 질문에 대한 답변과 함께 리뷰요청 보내겠습니다!

> 실패하는 경우를 인지한다는 의미는 예외케이스를 파악할 수 있다는 말일까요??

네 맞습니다! 실패하는 테스트 코드로 어떤 케이스에서 실패하는지 알 수 있을것 같아요! 혹시 다른 의미가 있을까요?

> 객체가 지니는 행동은 어떠한 기준으로 부여하고 분리하나요?

명확한 기준이 없어 사람마다 다르지만 외부에 제공해야 하는 기능을 수행할 가장 어울리는 객체를 선택합니다! 객체의 선택 기준은 해당 기능을 수행하기 위해서 얼마나 많은 정보를 알고있는가? 로 판단하는것 같아요! 

> 객체에게 어울리지 않은 행동은 어떻게 알 수 있을까요?

객체가 해당 행동을 수행하기에 적절한 정보나 상태가 없다면 해당 객체는 행동을 수행하기가 어색하다고 판단하는것 같습니다!

응집도는 무엇일까요?

`객체가 얼마나 자신의 역할을 잘 하고 있는가?`로 생각됩니다. 객체가 자신의 일을 외부에 알리지 않고 잘 수행한다면 응집도가 높다고 생각합니다!

> 고수준과 저수준은 무엇을 의미하나요?? 고수준 모듈이 저수준 모듈에 의존하면 안되는 이유는 무엇일까요? DTO의 변경이 도메인 객체에 영향을 주면 안되나요? DTO는 view의 영역인가요??

명확한 차이가 와닿지 않지만 고수준은 추상적인것이고 저수준은 그보다 구체화된 것 이라고 생각합니다. 구체는 변경이 일어날 가능성이 높기때문에 만약 추상이 구체에 의존하게 된다면 변경에 유연하게 대응할 수 없을거라 생각합니다. 
DTO에 변경이 일어났을때 도메인까지 변경이 일어나게 된다면 DTO가 변경될때 마다 도메인 객체를 수정해야하고, 프로그램 규모가 커지면 도메인과 DTO의 책임이 분리되지 않아 유지보수에 어려움이 생길 거라고 생각해요. 
DTO는 View의 영역보단 도메인과 View의 사이에 존재하는거라고 생각합니다. View에 도메인을 직접 넘겨서 사용하는것 보단 DTO를 한번 거쳐 필요한 데이터만 안전하게 넘길 수 있다고 생각해요!

리뷰 수고 많으셨습니다 : )

sinb57 reviewed (approved):
도기~ 고생 많으셨어요!! 2단계 진행을 위해 머지할게요ㅎ 고생 많으셨어요!! 2단계에서 봬요ㅎ 응원할게요~

---

>> 실패하는 경우를 인지한다는 의미는 예외케이스를 파악할 수 있다는 말일까요??
>
> 네 맞습니다! 실패하는 테스트 코드로 어떤 케이스에서 실패하는지 알 수 있을것 같아요! 혹시 다른 의미가 있을까요?

사람마다 다르겠지만 저는 테스트코드를 통해 "외부의 시점으로 바라보는" 연습을 많이 해요. 구현을 하면서 제공자의 시선으로 바라보는게 아니라, 테스트코드를 작성하면서 사용자의 관점에서 메서드 시그니처를 구상하는 등 메시징과 같이 의도를 드러내는 행위를 중요시 해요. 이러한 작업을 통해 객체의 역할과 책임을 잘 부여하며 설계를 이어갑니다.

> 아직 책임의 기준을 명확하게 세우는 것이 조금 힘든것 같아요. 이프는 무엇을 기준으로 객체를 분리하나요?

저는 책임을 변경에 연관지어 생각해요. 변경해야 하는 상황이 있을 때 변경의 범위를 기준으로 잘 묶여 있다면 응집도가 높다고 보고 있고, 이를 기반으로 객체를 설계해 나가요.

>> 고수준과 저수준은 무엇을 의미하나요?? 고수준 모듈이 저수준 모듈에 의존하면 안되는 이유는 무엇일까요? DTO의 변경이 도메인 객체에 영향을 주면 안되나요? DTO는 view의 영역인가요??
>
> 명확한 차이가 와닿지 않지만 고수준은 추상적인것이고 저수준은 그보다 구체화된 것 이라고 생각합니다. 구체는 변경이 일어날 가능성이 높기때문에 만약 추상이 구체에 의존하게 된다면 변경에 유연하게 대응할 수 없을거라 생각합니다.
DTO에 변경이 일어났을때 도메인까지 변경이 일어나게 된다면 DTO가 변경될때 마다 도메인 객체를 수정해야하고, 프로그램 규모가 커지면 도메인과 DTO의 책임이 분리되지 않아 유지보수에 어려움이 생길 거라고 생각해요.
DTO는 View의 영역보단 도메인과 View의 사이에 존재하는거라고 생각합니다. View에 도메인을 직접 넘겨서 사용하는것 보단 DTO를 한번 거쳐 필요한 데이터만 안전하게 넘길 수 있다고 생각해요!

고수준과 저수준은 "오브젝트" 책을 읽어보면 좋겠지만, 우선은 "객체지향의 사실과 오해"를 읽어보면 좋겠어요. 사실 DTO도 지금 시점에서 다루기엔 섣부르긴 해요. Level 2에서 본격적으로 다루게 될 내용이고, 지금은 DTO 없이도 구현 및 설계가 충분히 가능하거든요.

view가 도메인에 의존하면 안되는지, 왜 getter의 사용을 지양해야 하는지 등을 고민해보면 좋겠습니다.

그래도 본 PR에서 DTO에 대한 얘기가 마무리되지 않은 듯 하여 제 생각을 공유드리자면, DTO는 "계층"과 무관한 객체예요. 따라서 단순히 도메인과 view의 관계를 끊기 위해 사용하는 도구라고 보지 않으셨으면 좋겠어요. 오히려 DTO의 사용으로 getter의 사용이 많아진 것은 아닌지 의심해보셨으면 좋고, 객체지향으로의 사고를 억제하고 있는 것은 아닐지 돌이켜보셨으면 좋겠어요.
